clash-lib/prims/vhdl/Clash_Sized_Internal_BitVector.primitives

```diff
@@ -531,31 +531,57 @@ end process;
     , "kind"      : "Declaration"
     , "type"      : "shiftL# :: KnownNat n => BitVector n -> Int -> BitVector n"
     , "template"  :
-"~RESULT <= std_logic_vector(shift_left(unsigned(~ARG[1]), to_integer(~ARG[2])))
-    -- pragma translate_off
-    when (~ARG[2] >= 0) else (others => 'X')
-    -- pragma translate_on
-    ;"
+"~GENSYM[~RESULT_shiftL][0] : block
+  signal ~GENSYM[sh][1] : natural;
+begin
+  ~SYM[1] <=
+      -- pragma translate_off
+      natural'high when (~VAR[shI][2](~SIZE[~TYP[2]]-1 downto 31) /= 0) else
+      -- pragma translate_on
+      to_integer(~VAR[shI][2]
+      -- pragma translate_off
+      (30 downto 0)
+      -- pragma translate_on
+      );
+  ~RESULT <= std_logic_vector(shift_left(unsigned(~ARG[1]),~SYM[1]))
+      -- pragma translate_off
+      when (~ARG[2] >= 0) else (others => 'X')
+      -- pragma translate_on
+      ;
+end block;"
     }
   }
 , { "BlackBox" :
     { "name"      : "Clash.Sized.Internal.BitVector.shiftR#"
     , "kind"      : "Declaration"
     , "type"      : "shiftR# :: KnownNat n => BitVector n -> Int -> BitVector n"
     , "template"  :
-"~RESULT <= std_logic_vector(shift_right(unsigned(~ARG[1]),to_integer(~ARG[2])))
-    -- pragma translate_off
-    when (~ARG[2] >= 0) else (others => 'X')
-    -- pragma translate_on
-    ;"
+"~GENSYM[~RESULT_shiftR][0] : block
+  signal ~GENSYM[sh][1] : natural;
+begin
+  ~SYM[1] <=
+      -- pragma translate_off
+      natural'high when (~VAR[shI][2](~SIZE[~TYP[2]]-1 downto 31) /= 0) else
+      -- pragma translate_on
+      to_integer(~VAR[shI][2]
+      -- pragma translate_off
+      (30 downto 0)
+      -- pragma translate_on
+      );
+  ~RESULT <= std_logic_vector(shift_right(unsigned(~ARG[1]),~SYM[1]))
+      -- pragma translate_off
+      when (~ARG[2] >= 0) else (others => 'X')
+      -- pragma translate_on
+      ;
+end block;"
     }
   }
 , { "BlackBox" :
     { "name"      : "Clash.Sized.Internal.BitVector.rotateL#"
     , "kind"      : "Declaration"
     , "type"      : "rotateL# :: KnownNat n => BitVector n -> Int -> BitVector n"
     , "template"  :
-"~RESULT <= std_logic_vector(rotate_left(unsigned(~ARG[1]),to_integer(~ARG[2])))
+"~RESULT <= std_logic_vector(rotate_left(unsigned(~ARG[1]),to_integer(~VAR[sh][2](30 downto 0))))
```

Comment:
I'm worried that the masking for `rotate` only works properly for power-of-two bitlengths, when I define "works properly" as "simulation in Clash and in HDL give the same result for really large rotate amounts":

```haskell
>>> import qualified Prelude as P
>>> (1 :: BitVector 5) `rotateR` (2 P.^ 31 + 5)
0b0_0100
>>> (1 :: BitVector 5) `rotateR` ((2 P.^ 31 + 5) `mod` (2 P.^ 31))
0b0_0001
```

We should probably put "modulo bitlength" inside a `--pragma translate off` instead.

Reply:
Indeed, if I change `tests/shouldwork/Numbers/ShiftRotateBase.hs` to contain a really large amount, the testbench fails in GHDL:

```diff
@@ -5,12 +5,13 @@ import Clash.Prelude
 import Clash.Explicit.Testbench
 import Data.Word
 import Data.Int
+import qualified Prelude as P
 
 testpattern = 0b1010011000
 
 -- test with shift/rotate amounts: 0..16,200
 amounts :: Vec _ Int
-amounts = $(listToVecTH [0::Int ..16]) :< 200
+amounts = $(listToVecTH [0::Int ..16]) :< 200 :< (2 P.^ 31 + 5)
 
 testall v i
   = ( testAs @(Unsigned 8)  v i
```

Reply:
We can do:
```
#include "MachDeps.h"
```
Which gives us access to
```
WORD_SIZE_IN_BITS
```
So we can differentiate between 32 and 64 bit wrt the inputs we feed to the DUT and the outputs we expect from the DUT using CPP if.

Reply:
What code do you have in mind where knowing `WORD_SIZE_IN_BITS` would help?

On a 32-bit OS, the current code is a no-op and there was no issue to be fixed, the original code already worked fine.

On a 64-bit OS, the current code is broken, it only works for power-of-two data sizes.

If however we add a "modulo size" inside a `--pragma translate off`, that fixes the issue, it doesn't need `WORD_SIZE_IN_BITS`.

Reply:
I was referring to the Haskell side of things.